ash/src/vk.rs

```diff
@@ -82,6 +66,103 @@ pub(crate) unsafe fn ptr_chain_iter<T: ?Sized>(
     })
 }
 
+/// Structures implementing this trait are layout-compatible with [`BaseInStructure`] and
+/// [`BaseOutStructure`]. Such structures have an `s_type` field indicating its type, which must
+/// always match the value of [`TaggedStructure::STRUCTURE_TYPE`].
+pub unsafe trait AnyTaggedStructure<'a> {
+    /// Prepends the given extension struct between the root and the first pointer. This
+    /// method only exists on structs that can be passed to a function directly. Only
+    /// valid extension structs can be pushed into the chain.
+    /// If the chain looks like `A -> B -> C`, and you call `A.push(&mut D)`, then the
+    /// chain will look like `A -> D -> B -> C`.
+    ///
+    /// # Panics
+    /// If `next` contains a pointer chain of its own, this function will panic.  Call
+    /// `unsafe` [`Self::extend()`] to insert this chain instead.
+    fn push<T: Extends<'a, Self> + ?Sized>(mut self, next: &'a mut T) -> Self
```

Comment:
If the goal if this trait is object safety, do we need to revise this not to take `self` by value? You can't pass unsized types by value, after all.

Reply:
No, I think the first goal was to provide object safety for the `next` parameter `T: Extends<Self>` first. We can take a look at reworking the builder pattern to support that later, perhaps?